// mcp-server-monitor/dist/utils.js
import si from "systeminformation";
export function formatBytes(bytes) {
    if (bytes === 0)
        return "0 B";
    const k = 1024;
    const sizes = ["B", "KB", "MB", "GB", "TB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`;
}
export function formatUptime(seconds) {
    const days = Math.floor(seconds / 86400);
    const hours = Math.floor((seconds % 86400) / 3600);
    const mins = Math.floor((seconds % 3600) / 60);
    const secs = Math.floor(seconds % 60);
    const parts = [];
    if (days > 0)
        parts.push(`${days}d`);
    if (hours > 0)
        parts.push(`${hours}h`);
    if (mins > 0)
        parts.push(`${mins}m`);
    parts.push(`${secs}s`);
    return parts.join(" ");
}
export function formatPercent(value) {
    return `${value.toFixed(1)}%`;
}
export async function getSystemSummary() {
    const [cpu, mem, osInfo, time, load] = await Promise.all([
        si.cpu(),
        si.mem(),
        si.osInfo(),
        si.time(),
        si.currentLoad(),
    ]);
    return {
        os: `${osInfo.distro} ${osInfo.release} (${osInfo.arch})`,
        hostname: osInfo.hostname,
        platform: osInfo.platform,
        kernel: osInfo.kernel,
        uptime: formatUptime(time.uptime),
        cpu: `${cpu.manufacturer} ${cpu.brand} (${cpu.cores} cores, ${cpu.speedMax}GHz)`,
        cpuLoad: formatPercent(load.currentLoad),
        memoryTotal: formatBytes(mem.total),
        memoryUsed: formatBytes(mem.used),
        memoryFree: formatBytes(mem.free),
        memoryUsage: formatPercent((mem.used / mem.total) * 100),
    };
}
