// mcp-server-monitor/dist/tools/alerts.js
import { z } from "zod";
import si from "systeminformation";
import { formatBytes, formatPercent } from "../utils.js";
export function registerAlertTools(server) {
    server.tool("health_check", "Run a comprehensive server health check. Returns OK/WARNING/CRITICAL for CPU, memory, disk, and swap.", {
        cpuWarning: z.number().default(70).describe("CPU warning threshold (%)"),
        cpuCritical: z.number().default(90).describe("CPU critical threshold (%)"),
        memWarning: z.number().default(75).describe("Memory warning threshold (%)"),
        memCritical: z.number().default(90).describe("Memory critical threshold (%)"),
        diskWarning: z.number().default(80).describe("Disk warning threshold (%)"),
        diskCritical: z.number().default(95).describe("Disk critical threshold (%)"),
    }, async ({ cpuWarning, cpuCritical, memWarning, memCritical, diskWarning, diskCritical }) => {
        const [load, mem, fs] = await Promise.all([
            si.currentLoad(),
            si.mem(),
            si.fsSize(),
        ]);
        const alerts = [];
        // CPU check
        const cpuLoad = load.currentLoad;
        alerts.push({
            level: cpuLoad >= cpuCritical ? "critical" : cpuLoad >= cpuWarning ? "warning" : "ok",
            metric: "CPU Load",
            value: formatPercent(cpuLoad),
            threshold: `warn: ${cpuWarning}% / crit: ${cpuCritical}%`,
            message: cpuLoad >= cpuCritical ? "CPU critically high!" : cpuLoad >= cpuWarning ? "CPU load elevated" : "CPU load normal",
        });
        // Memory check
        const memPct = (mem.used / mem.total) * 100;
        alerts.push({
            level: memPct >= memCritical ? "critical" : memPct >= memWarning ? "warning" : "ok",
            metric: "Memory",
            value: `${formatPercent(memPct)} (${formatBytes(mem.used)}/${formatBytes(mem.total)})`,
            threshold: `warn: ${memWarning}% / crit: ${memCritical}%`,
            message: memPct >= memCritical ? "Memory critically high!" : memPct >= memWarning ? "Memory usage elevated" : "Memory usage normal",
        });
        // Swap check
        if (mem.swaptotal > 0) {
            const swapPct = (mem.swapused / mem.swaptotal) * 100;
            alerts.push({
                level: swapPct >= 80 ? "critical" : swapPct >= 50 ? "warning" : "ok",
                metric: "Swap",
                value: `${formatPercent(swapPct)} (${formatBytes(mem.swapused)}/${formatBytes(mem.swaptotal)})`,
                threshold: "warn: 50% / crit: 80%",
                message: swapPct >= 80 ? "Swap critically high!" : swapPct >= 50 ? "Swap usage elevated" : "Swap usage normal",
            });
        }
        // Disk checks
        for (const d of fs) {
            const pct = d.use;
            alerts.push({
                level: pct >= diskCritical ? "critical" : pct >= diskWarning ? "warning" : "ok",
                metric: `Disk (${d.mount})`,
                value: `${formatPercent(pct)} (${formatBytes(d.used)}/${formatBytes(d.size)})`,
                threshold: `warn: ${diskWarning}% / crit: ${diskCritical}%`,
                message: pct >= diskCritical ? `Disk ${d.mount} critically full!` : pct >= diskWarning ? `Disk ${d.mount} filling up` : `Disk ${d.mount} OK`,
            });
        }
        const icons = { ok: "✅", warning: "⚠️", critical: "🚨" };
        const overallLevel = alerts.some((a) => a.level === "critical")
            ? "critical"
            : alerts.some((a) => a.level === "warning")
                ? "warning"
                : "ok";
        const lines = [
            `${icons[overallLevel]}  Health Check — ${overallLevel.toUpperCase()}`,
            `═══════════════════════════════════════════════════════`,
        ];
        for (const a of alerts) {
            lines.push(`${icons[a.level]} ${a.metric}: ${a.value} — ${a.message}`);
        }
        lines.push(``, `Thresholds used:`);
        for (const a of alerts) {
            lines.push(`  ${a.metric}: ${a.threshold}`);
        }
        return { content: [{ type: "text", text: lines.join("\n") }] };
    });
    server.tool("performance_report", "Generate a full performance report combining CPU, memory, disk, network, and top processes.", {}, async () => {
        const [load, mem, fs, net, procs, osInfo, time] = await Promise.all([
            si.currentLoad(),
            si.mem(),
            si.fsSize(),
            si.networkStats(),
            si.processes(),
            si.osInfo(),
            si.time(),
        ]);
        const topCpu = procs.list.sort((a, b) => b.cpu - a.cpu).slice(0, 5);
        const topMem = procs.list.sort((a, b) => b.mem - a.mem).slice(0, 5);
        const memPct = (mem.used / mem.total) * 100;
        const lines = [
            `📊  Performance Report`,
            `═══════════════════════════════════════════════════════`,
            `Generated: ${new Date().toISOString()}`,
            `Host:      ${osInfo.hostname} (${osInfo.distro} ${osInfo.release})`,
            `Uptime:    ${Math.floor(time.uptime / 86400)}d ${Math.floor((time.uptime % 86400) / 3600)}h ${Math.floor((time.uptime % 3600) / 60)}m`,
            ``,
            `── CPU ──────────────────────────────────────────────`,
            `Load: ${formatPercent(load.currentLoad)} (user: ${formatPercent(load.currentLoadUser)}, sys: ${formatPercent(load.currentLoadSystem)})`,
            ``,
            `── Memory ──────────────────────────────────────────`,
            `RAM:  ${formatBytes(mem.used)} / ${formatBytes(mem.total)} (${formatPercent(memPct)})`,
            `Swap: ${formatBytes(mem.swapused)} / ${formatBytes(mem.swaptotal)}`,
            ``,
            `── Disk ────────────────────────────────────────────`,
        ];
        for (const d of fs) {
            lines.push(`${d.mount}: ${formatBytes(d.used)}/${formatBytes(d.size)} (${formatPercent(d.use)})`);
        }
        lines.push(``, `── Network ─────────────────────────────────────────`);
        for (const n of net) {
            lines.push(`${n.iface}: ↓${formatBytes(n.rx_sec || 0)}/s  ↑${formatBytes(n.tx_sec || 0)}/s`);
        }
        lines.push(``, `── Top 5 by CPU ────────────────────────────────────`);
        for (const p of topCpu) {
            lines.push(`  ${p.name.padEnd(25)} ${formatPercent(p.cpu).padEnd(8)} PID ${p.pid}`);
        }
        lines.push(``, `── Top 5 by Memory ─────────────────────────────────`);
        for (const p of topMem) {
            lines.push(`  ${p.name.padEnd(25)} ${formatPercent(p.mem).padEnd(8)} PID ${p.pid}`);
        }
        lines.push(``, `── Process Summary ─────────────────────────────────`, `Total: ${procs.all}  Running: ${procs.running}  Sleeping: ${procs.sleeping}  Blocked: ${procs.blocked}`);
        return { content: [{ type: "text", text: lines.join("\n") }] };
    });
}
